Define and obtain the "Typst version" in a central location

har7an opened the pull request:
As discussed in #7336, here is my proposal to unifying **the** "Typst version" across all subcrates. I've split it into a commit per component so you can trace the change more easily (I hope). To the best of my knowledge, there is no mention of `env!("CARGO_PKG_VERSION*")` or `env!("TYPST_VERSION")` in the code any more beyond what `TypstVersion` reads.

I went ahead and added the global `typst::version()` function as proposed in #7336 (see 07f49595) and the subsequent commit (see 567dee88) builds on top of that. I can change that to use the version definition from `typst-syntax`, though.

I chose to implement the global Typst version in `typst-syntax` simply because it was the most "low-level" crate that read out the Typst version from the environment. If you feel there is a better home for my code, please let me know. I also haven't written any tests because I'm not sure what (and how) to test. If you have ideas, I'm open to that as well.


## User-facing changes

**This change is invisible to Typst documents.**

It only affects Rust developers. Anyone using the `typst-syntax` or `typst` crate can now query the Typst version they are using (with `typst_syntax::TypstVersion::new()` or `typst::version()` respectively). This affects for example <https://github.com/typst/package-check/pull/37>. That is of course only true in case 07f49595 is allowed to stay.

ping @laurmaedje 

Closes: #7336

Andrew15-5 commented:
The commits are usually squashed, so there are no restrictions on commit history manipulations. You don't have to rebase.

Andrew15-5 commented on crates/typst-syntax/build.rs:
```diff
@@ -0,0 +1,7 @@
+
+    if option_env!("TYPST_VERSION").is_none() {
+        println!("cargo:rustc-env=TYPST_VERSION={}", env!("CARGO_PKG_VERSION"));
+    }
```
Why redefine `TYPST_VERSION` if `CARGO_PKG_VERSION` is fetched separately? It also might be invalid, so having an up-to-date backup is important.

```suggestion
    println!("cargo:rerun-if-env-changed=TYPST_VERSION");
    println!("cargo:rerun-if-env-changed=CARGO_PKG_VERSION");
```

har7an replied:
I don't know about *backup*. If I were a packager and a project gave me an env variable (`TYPST_VERSION` in this case) to "override" what the package reports as its version, I'd like this mechanism to "work" or at least break properly.

The way I understand your proposal, you try to parse `TYPST_VERSION` and fall back to `CARGO_PKG_VERSION` if the former doesn't exist **or** doesn't parse into a valid semver object. As a package maintainer, I'd expect the code to explode if my input (to `TYPST_VERSION`) is invalid, rather than silently having it overwritten.

Or did I misunderstand your code?

Andrew15-5 replied:
Considering committed behavior, why check for `CARGO_PKG_VERSION` [here](https://github.com/typst/typst/blob/a73515551ccddafb18090b3fa1c0785a02248b7a/crates/typst-syntax/src/version.rs#L50) at all, if at compile time `TYPST_VERSION` is **already** equal to `CARGO_PKG_VERSION`? Makes no sense.

`build.rs` and `TypstVersion::new()` must correctly negotiate env vars and their values to work as expected. One depends on the other.

Andrew15-5 commented on crates/typst-syntax/src/package.rs:
```diff
@@ -478,6 +475,32 @@ impl<'de> Deserialize<'de> for PackageVersion {
+            patch: digit!(patch),
+        })
+    }
+}
```
When will this return `Err` variant?

har7an commented:
I think I fixed all of your feedback now. I'll reply on all your comments here because GitHub apparently offers no way to batch replies and I don't want to spam into your Inbox...

> Considering committed behavior, why check for CARGO_PKG_VERSION [here](https://github.com/typst/typst/blob/a73515551ccddafb18090b3fa1c0785a02248b7a/crates/typst-syntax/src/version.rs#L50) at all, if at compile time TYPST_VERSION is already equal to CARGO_PKG_VERSION? Makes no sense.
>
> build.rs and TypstVersion::new() must correctly negotiate env vars and their values to work as expected. One depends on the other.

True, I adopted your proposal for the `build.rs` now. If both `TYPST_VERSION` and `CARGO_PKG_VERSION` are defined, the code will prioritize `TYPST_VERSION`. If that in turn has an invalid version number, `CARGO_PKG_VERSION` is *not* checked and the code simply explodes.

> When will this return Err variant?

It should have all along.

And regarding my own comment:

> Agreed, I'm looking into this right now. I initially chose to impl Display due to [this function in the Error trait](https://doc.rust-lang.org/stable/std/error/trait.Error.html#method.description) but apparently that doesn't work as I expected it to.

I tend to work with `thiserror` for defining error types, hence the mistake. `unwrap` and `expect` both call an [internal function `unwrap_failed`][1] eventually, and that displays the `Err` object with its `Debug` impl. I fixed this now by replacing the `Display` impl with `Debug` (because of KISS and such).

[1]: https://doc.rust-lang.org/1.91.1/src/core/result.rs.html#1854

Andrew15-5 commented:
Oh, neat. That's a cool macro. Oh, you've created it, I thought it's in std, should've looked at the one commit change. Um, I'm not sure the additional 18 lines are worth it over keeping the slightly more verbose pattern matching.

laurmaedje commented:
Thanks for the PR! Unfortunately, I must say that this seems a bit over-engineered overall (also indicated by the relatively large additive diff). 

For instance, the `VersionError` seems unnecessary as it will be immediately unwrapped and `EnvWithSource` also seems a bit much. From the `TryFrom` impls both uses are immediately unwrapped, so the complex error handling and macros there have limited benefit. (This conversion will never panic in practice and if it does, the stack trace is enough.)

As for the placement, I think `typst_utils` would be a better place for this than `typst_syntax`.

As-is, the refactor does not seem worth it to me, but I think this could be simplified down into a worthwhile change.

har7an commented:
Thanks for the thorough review! I'm not sure what to do about your remaining comments though.

> this could use crate::singleton!.

That macro doesn't make sense as I understand it, but I'm not familiar with `LazyLock` so I'll believe you on this. I added tests to make sure and they look alright.

> I think I'd directly put the major/minor/patch triple here (essentially only using semver for extracing major/minor/patch) and then we can make the type Copy and return it by value everywhere.

I feel this contradicts the idea of having a lock to guarantee a singleton object instance. Why make a singleton if we allow consumers to copy it?

I implemented both the `singleton!` macro and `Copy` trait now. Is this what you wanted?

Andrew15-5 commented on crates/typst-utils/src/version.rs:
```diff
@@ -0,0 +1,90 @@
+/// This structure contains the current Typst version. To query the precise
+/// version number, refer to the [`TypstVersion::major()`],
+/// [`TypstVersion::minor()`] and [`TypstVersion::patch()`] functions. You can
+/// read the underlying, raw version string (e.g. for CLI output) with
```
```suggestion
/// read the underlying, raw version string (e.g., for CLI output) with
```

Andrew15-5 replied:
This one was not addressed.

Andrew15-5 commented on crates/typst-utils/src/version.rs:
```diff
@@ -0,0 +1,90 @@
+/// [`TypstVersion::raw`].
+///
+/// Optionally, this may also contain the Git commit from which Typst was built.
+/// However, this field may be unpopulated.
```
It doesn't contain the commit itself, but the hash value of it.

```suggestion
/// Optionally, this may also contain the hash value of the Git commit from
/// which Typst was built. However, this field may be unpopulated.
```

Andrew15-5 commented on crates/typst-utils/src/version.rs:
```diff
@@ -0,0 +1,90 @@
+    patch: u32,
+    /// Raw, unmodified version string.
+    raw: &'static str,
+    /// The commit SHA.
```
"Commit hash" is also fine.

Also, here it should be uppercased or changed to other term:

https://github.com/typst/typst/blob/2c569ad759ab9624b236a94bf602634f0b4ce941/crates/typst-utils/src/version.rs#L11

Andrew15-5 commented on crates/typst-utils/src/version.rs:
```diff
@@ -7,8 +7,10 @@
-/// time. A fallback mechanism is currently implemented in `build.rs`.
+/// Build tooling can set these environment variables to configure the exposed
+/// information. If the environment variables are left unset, the values are
+/// populated via `build.rs` from the Cargo package manifest version and the git
```
Git should be capitalized.

laurmaedje commented on crates/typst-utils/src/version.rs:
```diff
@@ -0,0 +1,99 @@
+    }
+}
+
+/// Displays the commit Typst was built from human-readably.
```
There wasn't a clean way to make use of it in `typst info` otherwise. You can try and see for yourself.
